False positive for `trailing_closure` rule with anonymous closure call

### New Issue Checklist

- [✅] Updated SwiftLint to the latest version
- [✅] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Bug Report

It is sometimes handy to use anonymous closure as code grouping, e.g.:

```swift
menu.addItem({
    let menuItem = NSMenuItem(title: "Blah!")
    return menuItem
}())
```

When linted with `trailing_closure` enable, this incorrectly triggers a warning.

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
Bagombo:closure proxi$ swiftlint 
Loading configuration from '.swiftlint.yml'
Linting Swift files in current working directory
Linting 'Closure.swift' (1/1)
Closure.swift:2:5: warning: Trailing Closure Violation: Trailing closure syntax should be used whenever possible. (trailing_closure)
Done linting! Found 1 violation, 0 serious in 1 file.
Bagombo:closure proxi$ 
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.25.1
* Installation method used (Homebrew, CocoaPods, building from source, etc)? CocoaPods
* Paste your configuration file:

```yml
opt_in_rules:
    - trailing_closure
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  If so, paste their relative paths and respective contents. No
* Which Xcode version are you using (check `xcode-select -p`)? 9.3
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

```swift
// This triggers a violation:
func foo(_ bar: Int) {
    foo({ return 42 }())
}
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -14,7 +14,10 @@
 
 #### Bug Fixes
 
-* None.
+* Fix false positives on `trailing_closure` rule when using anonymous closure
+  calls.  
+  [Marcelo Fabri](https://github.com/marcelofabri)
+  [#2159](https://github.com/realm/SwiftLint/issues/2159)
 
 ## 0.30.0: A New Washer and Dryer Set
 

--- a/Rules.md
+++ b/Rules.md
@@ -17785,6 +17785,18 @@ offsets.sorted { $0.offset < $1.offset }
 
 ```
 
+```swift
+foo.something({ return 1 }())
+```
+
+```swift
+foo.something({ return $0 }(1))
+```
+
+```swift
+foo.something(0, { return 1 }())
+```
+
 </details>
 <details>
 <summary>Triggering Examples</summary>

--- a/Source/SwiftLintFramework/Rules/Style/TrailingClosureRule.swift
+++ b/Source/SwiftLintFramework/Rules/Style/TrailingClosureRule.swift
@@ -17,7 +17,10 @@ public struct TrailingClosureRule: OptInRule, ConfigurationProviderRule {
             "foo.reduce(0) { $0 + 1 }\n",
             "if let foo = bar.map({ $0 + 1 }) { }\n",
             "foo.something(param1: { $0 }, param2: { $0 + 1 })\n",
-            "offsets.sorted { $0.offset < $1.offset }\n"
+            "offsets.sorted { $0.offset < $1.offset }\n",
+            "foo.something({ return 1 }())",
+            "foo.something({ return $0 }(1))",
+            "foo.something(0, { return 1 }())"
         ],
         triggeringExamples: [
             "↓foo.map({ $0 + 1 })\n",
@@ -63,8 +66,9 @@ public struct TrailingClosureRule: OptInRule, ConfigurationProviderRule {
     }
 
     private func shouldBeTrailingClosure(dictionary: [String: SourceKitRepresentable], file: File) -> Bool {
-        func isTrailingClosure() -> Bool {
-            return isAlreadyTrailingClosure(dictionary: dictionary, file: file)
+        func shouldTrigger() -> Bool {
+            return !isAlreadyTrailingClosure(dictionary: dictionary, file: file) &&
+                !isAnonymousClosureCall(dictionary: dictionary, file: file)
         }
 
         let arguments = dictionary.enclosedArguments
@@ -74,7 +78,7 @@ public struct TrailingClosureRule: OptInRule, ConfigurationProviderRule {
             case let closureArguments = filterClosureArguments(arguments, file: file),
             closureArguments.count == 1,
             closureArguments.last?.bridge() == arguments.last?.bridge() {
-            return !isTrailingClosure()
+            return shouldTrigger()
         }
 
         // check if there's only one unnamed parameter that is a closure
@@ -88,7 +92,7 @@ public struct TrailingClosureRule: OptInRule, ConfigurationProviderRule {
             let range = file.contents.bridge().byteRangeToNSRange(start: start, length: length),
             let match = regex("\\s*\\(\\s*\\{").firstMatch(in: file.contents, options: [], range: range)?.range,
             match.location == range.location {
-            return !isTrailingClosure()
+            return shouldTrigger()
         }
 
         return false
@@ -117,5 +121,17 @@ public struct TrailingClosureRule: OptInRule, ConfigurationProviderRule {
         }
 
         return !text.hasSuffix(")")
+    }
+
+    private func isAnonymousClosureCall(dictionary: [String: SourceKitRepresentable],
+                                        file: File) -> Bool {
+        guard let offset = dictionary.offset,
+            let length = dictionary.length,
+            let range = file.contents.bridge().byteRangeToNSRange(start: offset, length: length) else {
+                return false
+        }
+
+        let pattern = regex("\\)\\s*\\)\\z")
+        return pattern.numberOfMatches(in: file.contents, range: range) > 0
     }
 }

